### PR TITLE
feat: added api testing

### PR DESCRIPTION
**Pull request by Aleksey28**
## Purpose
Implement API testing

## Approach
- [x] Add interfaces
- [x] Add building `Request` instance and export it
- [x] Decorate `Request` with additional method `get`, `post`, `delete`, `put`, `patch`, `head`
- [x] Valiate options
- [x] Cover a request function with `ReExecutablePromise`
- [x] Add a client command for the proxing url
- [x] Prepare and proxy an url
- [x] Transform a body according the headers
- [x] Create an `externalProxySettings`
- [x] Add creating request options via `RequestOptions` class from HH
- [x] Send request with `DestinationRequest`
- [x] Process response body
- [x] Add exchange cookies
- [x] Add possibility to set timeout
- [x] Add possibility to specify sending credentials
- [x] Add possibility to set auth option
- [x] Add possibility to set proxy
- [x] Add possibility to specify response (raw or not)
- [x] Add possibility to send relative request from the current url

```ts
type Params = URLSearchParams | {[key: string]: string | number | boolean}
type ResponseBody = IncomingMessage | Buffer | object | string;

interface AuthOptions {
    username: string;
    password: string;
}

interface ProxyOptions {
    protocol?: string;
    host: string;
    port: number | string;
    auth?: AuthOptions;
}

interface RequestOptions {
    /**
     *  Request url.
     */
    url?: string | URL;
    /**
     * Request methods.
     */
    method?: string;
    /**
     * Request headers.
     */
    headers?: Record<string, string>;
    /**
     * Request params.
     */
    params?: Params;
    /**
     * Request body.
     */
    body?: object;
    /**
     * Request timeout.
     */
    timeout?: number;
    /**
     * Request credentials.
     */
    withCredentials?: boolean;
    /**
     * Request auth.
     */
    auth?: AuthOptions;
    /**
     * Request proxy.
     */
    proxy?: ProxyOptions;
    /**
     * Return raw response.
     */
    rawResponse?: boolean;
}

interface ResponseOptions {
    /**
     * Response status.
     */
    status: number;
    /**
     * Response status text.
     */
    statusText: string;
    /**
     * Response headers.
     */
    headers: object;
    /**
     * Response body.
     */
    body: ResponseBody;
}

type RequestPromise = RequestAPI | Promise<ResponseOptions>
type RequestUrlOpts = string | URL | RequestOptions;

t.request(urlOpts: RequestUrlOpts, options?: RequestOptions): RequestPromise
t.request.get(urlOpts: RequestUrlOpts, options?: RequestOptions): RequestPromise
t.request.delete(urlOpts: RequestUrlOpts, options?: RequestOptions): RequestPromise
t.request.head(urlOpts: RequestUrlOpts, options?: RequestOptions): RequestPromise
t.request.post(urlOpts: RequestUrlOpts, options?: RequestOptions): RequestPromise
t.request.put(urlOpts: RequestUrlOpts, options?: RequestOptions): RequestPromise
t.request.patch(urlOpts: RequestUrlOpts, options?: RequestOptions): RequestPromise
```

## References
https://github.com/DevExpress/testcafe-hammerhead/pull/2758
Closes DevExpress/testcafe-private#1

## Pre-Merge TODO
- [x] Write tests for your proposed changes
- [x] Make sure that existing tests do not fail

### CODE REVIEW

File: src/request/request-builder.ts [hunk 1]
```diff
@@ -0,0 +1,142 @@
+import testRunTracker from '../api/test-run-tracker';
```
Comment (miherlosev) [inline src/request/request-builder.ts:1]: We don't need an additional library for sending request.
Use the [DestinationRequest](https://github.com/DevExpress/testcafe-hammerhead/blob/master/src/request-pipeline/destination-request/index.ts) from hammerhead.

**Review by miherlosev — CHANGES_REQUESTED** [review]

File: package.json [hunk 2]
```diff
@@ -75,8 +75,10 @@
     "@babel/preset-react": "^7.12.1",
     "@babel/runtime": "^7.12.5",
     "@miherlosev/esm": "3.2.26",
+    "@types/axios": "^0.14.0",
     "@types/node": "^12.20.10",
     "async-exit-hook": "^1.1.2",
+    "axios": "^0.26.1",
```
Comment (miherlosev) [inline package.json:81]: Remove it

**Review by miherlosev — CHANGES_REQUESTED** [review]

File: src/request/dispatchRequest.ts [hunk 3]
```diff
@@ -0,0 +1,31 @@
+import TestRun from '../test-run';
```
Comment (miherlosev) [inline src/request/dispatchRequest.ts:1]: The `dispatch` word is usually used for events. Rename this file to `src/request/send`.

Remove the `request` word from all related file names. Class names in it should not be changed.

File: src/request/dispatchRequest.ts [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+import { DestinationRequest } from 'testcafe-hammerhead';
+import { IncomingMessage } from 'http';
+import { ExternalRequestOptions, ResponseOptions } from './interfaces';
+import { processRequestOptions } from './process-request-options';
+import { processResponseData } from './process-response-data';
+
+const DEFAULT_STATUS      = 404;
```
Comment (miherlosev) [inline src/request/dispatchRequest.ts:8]: Could you please add a test for this case?

File: src/request/interfaces.ts [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+    url: string | URL;
+    method?: Method;
+    headers?: OutgoingHttpHeaders;
+    params?: Params;
+    body?: object;
+    timeout?: number;
+    withCredentials?: boolean;
+    auth?: AuthOptions;
+    // NOTE: Isn't implemented
+    // maxRedirects?: number;
```
Comment (miherlosev) [inline src/request/interfaces.ts:38]: I propose to postpone the `maxRedirects` property until we the get customer's feedback.

File: src/request/process-request-options.ts [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+import CONTENT_TYPES from '../assets/content-types';
+import HTTP_HEADERS from '../utils/http-headers';
+import { RUNTIME_ERRORS } from '../errors/types';
+import { APIError } from '../errors/runtime';
+import { GetAjaxProxyUrlCommand } from '../test-run/commands/actions';
+
+const DEFAULT_ACCEPT            = { [HTTP_HEADERS.accept]: `${CONTENT_TYPES.json}, ${CONTENT_TYPES.textPlain}, ${CONTENT_TYPES.all}` };
+const DEFAULT_IS_REQUEST        = { [HTTP_HEADERS.isRequest]: true };
+const METHODS_WITH_CONTENT_TYPE = ['post', 'put', 'patch'];
+const DEFAULT_REQUEST_TIMEOUT   = 2 * 60 * 1000;
```
Comment (miherlosev) [inline src/request/process-request-options.ts:30]: It's not necessary. The default values are in the [hammerhead](https://github.com/DevExpress/testcafe-hammerhead/blob/master/src/request-pipeline/destination-request/default-request-timeout.ts) module.

File: src/request/process-request-options.ts [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+            castArray(params[key]).forEach(v => {
+                searchParams.append(key, typeof v === 'object' ? JSON.stringify(v) : String(v));
+            });
+        }
+    }
+
+    return `${url}${(url.includes('?') ? '&' : '?')}${searchParams.toString()}`;
+}
+
+export async function processRequestOptions (testRun: TestRun, options: ExternalRequestOptions, callsite: string): Promise<RequestOptions> {
```
Comment (miherlosev) [inline src/request/process-request-options.ts:163]: `createRequestOptions`

File: src/request/process-response-data.ts [hunk 8]
```diff
@@ -0,0 +1,39 @@
+import { Stream } from 'stream';
+import { Buffer } from 'buffer';
+import { ResponseOptions } from './interfaces';
+import { IncomingMessage } from 'http';
+import HTTP_HEADERS from '../utils/http-headers';
+import CONTENT_TYPES from '../assets/content-types';
+
+function streamToBuffer (stream: Stream): Promise<Buffer> {
```
Comment (miherlosev) [inline src/request/process-response-data.ts:8]: Use function from [hammerhead](https://github.com/DevExpress/testcafe-hammerhead/blob/master/src/utils/promisify-stream.ts).

**Review by miherlosev — CHANGES_REQUESTED** [review]

File: src/request/process-request-options.ts (hunk 6 above)
Comment (Aleksey28) [reply src/request/process-request-options.ts:30]: In the hammerhead, it is used only for creating sessions. I added using these values also for creating RequesOptions

File: src/test-run/commands/options.d.ts [hunk 9]
```diff
@@ -82,3 +82,10 @@ export class CookieOptions {
     public httpOnly: boolean;
     public sameSite: string;
 }
+
+// GetProxyUrl
```
Comment (miherlosev) [inline src/test-run/commands/options.d.ts:86]: Remove this comment here and below

File: ts-defs-src/test-api/request.d.ts [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+     */
+    auth?: AuthOptions;
+    /**
+     * Request proxy.
+     */
+    proxy?: ProxyOptions;
+    /**
+     * Process response.
+     */
+    processResponse?: boolean;
```
Comment (miherlosev) [inline ts-defs-src/test-api/request.d.ts:68]: Need to discuss this option name. I propose the `rawResponse` name. Contact with @titerman .

File: ts-defs-src/test-api/request.d.ts [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+     */
+    statusText: Promise<string>;
+    /**
+     * Response headers.
+     */
+    headers: Promise<object>;
+    /**
+     * Response body.
+     */
+    body: Promise<unknown>;
```
Comment (miherlosev) [inline ts-defs-src/test-api/request.d.ts:110]: Use more strict typing here.

File: ts-defs-src/test-api/request.d.ts [hunk 12]
```diff
@@ -0,0 +1,166 @@
...
+     */
+    statusText: string;
+    /**
+     * Response headers.
+     */
+    headers: object;
+    /**
+     * Response body.
+     */
+    body: unknown;
```
Comment (miherlosev) [inline ts-defs-src/test-api/request.d.ts:91]: Use more strict typing here.

File: test/functional/fixtures/request/testcafe-fixtures/request-test.js [hunk 13]
```diff
@@ -0,0 +1,316 @@
...
+('Should execute a request with relative url', async (t) => {
+    const { body } = await Request.get('/api/data');
+
+    await t.expect(body.data).eql({
+        name:     'John Hearts',
+        position: 'CTO',
+    });
+});
+
+//TODO: added tests:
```
Comment (miherlosev) [inline test/functional/fixtures/request/testcafe-fixtures/request-test.js:313]: Remove these comments.

**Review by miherlosev — CHANGES_REQUESTED** [review]
In general, it's excellent.

File: src/request/create-request-options.ts [hunk 14]
```diff
@@ -0,0 +1,214 @@
...
+import { RUNTIME_ERRORS } from '../errors/types';
+import { APIError } from '../errors/runtime';
+import { GetProxyUrlCommand } from '../test-run/commands/actions';
+
+const DEFAULT_ACCEPT            = { [HTTP_HEADERS.accept]: `${CONTENT_TYPES.json}, ${CONTENT_TYPES.textPlain}, ${CONTENT_TYPES.all}` };
+const DEFAULT_IS_REQUEST        = { [HTTP_HEADERS.isRequest]: true };
+const METHODS_WITH_CONTENT_TYPE = ['post', 'put', 'patch'];
+const DEFAULT_REQUEST_METHOD    = 'GET';
+
+function setContentTypeIfUnset (headers: OutgoingHttpHeaders, value: string): void {
```
Comment (miherlosev) [inline src/request/create-request-options.ts:32]: `set if unset` is a tautology. Rename to `setContentTypeIfNotExists`

File: src/request/create-request-options.ts [hunk 15]
```diff
@@ -0,0 +1,214 @@
...
+        isBuffer(body) ||
+        isStream(body)
+    )
+        return Buffer.from(body);
+    else if (ArrayBuffer.isView(body))
+        return Buffer.from(body.buffer);
+
+    else if (body instanceof URLSearchParams) {
+        setContentTypeIfUnset(headers, `${CONTENT_TYPES.urlencoded};charset=utf-8`);
+        return Buffer.from(body.toString());
```
Comment (miherlosev) [inline src/request/create-request-options.ts:64]: Add a new line before the `return` statement.

File: src/request/create-request-options.ts [hunk 16]
```diff
@@ -0,0 +1,214 @@
...
+        port:                  url.port,
+        path:                  prepareSearchParams(url.pathname + url.search, options.params),
+        auth:                  auth ? `${auth.username}:${auth.password}` : void 0,
+        headers:               headers,
+        externalProxySettings: externalProxySettings,
+        credentials:           testRun.session.getAuthCredentials(),
+        body:                  body,
+        rawHeaders:            void 0,
+        requestTimeout:        {
+            ajax: options.timeout,
```
Comment (miherlosev) [inline src/request/create-request-options.ts:207]: We should provide timeouts for both request types (`ajax` and `regular`).

File: src/request/create-request-options.ts [hunk 17]
```diff
@@ -0,0 +1,214 @@
...
+        },
+        disableHttp2: testRun.session.isHttp2Disabled(),
+        isAjax:       true,
```
Comment (miherlosev) [inline src/request/create-request-options.ts:210]: The sent request should use the `options.isAjax` property.

**Review by miherlosev — CHANGES_REQUESTED** [review]

**Review by miherlosev — APPROVED** [review]
Awesome

File: package.json [hunk 18]
```diff
@@ -137,7 +137,7 @@
     "source-map-support": "^0.5.16",
     "strip-bom": "^2.0.0",
     "testcafe-browser-tools": "2.0.23",
-    "testcafe-hammerhead": "24.5.20",
+    "testcafe-hammerhead": "24.5.21",
```
Comment (AndreyBelym) [inline package.json:140]: Pls rebase :(

**Review by AndreyBelym — APPROVED** [review]

**Pull request merged**